cannon/mipsevm/memory/binary_tree.go

```diff
@@ -0,0 +1,121 @@
+package memory
+
+import (
+	"math/bits"
+)
+
+// BinaryTreeIndex is a representation of the state of the memory in a binary merkle tree.
+type BinaryTreeIndex struct {
+	// generalized index -> merkle root or nil if invalidated
+	nodes map[uint64]*[32]byte
+	// Reference to the page table from Memory.
+	pageTable map[Word]*CachedPage
+}
+
+func NewBinaryTreeMemory() *Memory {
+	pages := make(map[Word]*CachedPage)
+	index := NewBinaryTreeIndex()
+	index.setPageBacking(pages)
```

Comment:
It would probably be better to have this initial call to `setPageBacking` done by th eNewBinaryTreeIndex` constructor so that it always returns a fully formed instance.